middle_end/compilenv.ml

```diff
@@ -104,8 +104,17 @@ let symbolname_for_pack pack name =
 
 let unit_id_from_name name = Ident.create_persistent name
 
+let no_leading_or_consecutive_underscores str =
+  String.mapi
+    (fun i c -> if c = '_' && (i = 0 || str.[i - 1] = '_') then '.' else c)
+    str
+
 let concat_symbol unitname id =
-  unitname ^ "__" ^ id
+  (* PR#8998: no_leading_or_consecutive_underscores prevents any
+     occurrence of "__" after the "__" we add, so concat_symbol is
```

Comment:
I would rather say: guarantees that the last occurrence of `__` in the symbol is the one we add. This covers both what you say (which I understand as talking about occurrences of `__` in `id`) and the reason why an initial `_` is better escaped.